pkg/manager/infosync/info.go

```diff
@@ -0,0 +1,209 @@
+	etcdCli         *clientv3.Client
+	wg              waitgroup.WaitGroup
+	cancelFunc      context.CancelFunc
+	topologySession *concurrency.Session
+}
+
+type syncConfig struct {
+	sessionTTL    int
+	refreshIntvl  time.Duration
+	putTimeout    time.Duration
+	putRetryIntvl time.Duration
+	putRetryCnt   uint64
+}
+
+type TopologyInfo struct {
+	Version        string `json:"version"`
+	GitHash        string `json:"git_hash"`
+	IP             string `json:"ip"`
+	Port           string `json:"port"`
+	StatusPort     string `json:"status_port"`
+	DeployPath     string `json:"deploy_path"`
+	StartTimestamp int64  `json:"start_timestamp"`
+}
+
+func NewInfoSyncer(etcdCli *clientv3.Client, lg *zap.Logger) *InfoSyncer {
+	return &InfoSyncer{
+		etcdCli: etcdCli,
+		lg:      lg,
+		syncConfig: syncConfig{
+			sessionTTL:    topologySessionTTL,
+			refreshIntvl:  topologyRefreshIntvl,
+			putTimeout:    topologyPutTimeout,
+			putRetryIntvl: topologyPutRetryIntvl,
+			putRetryCnt:   topologyPutRetryCnt,
+		},
+	}
+}
+
+func (is *InfoSyncer) Init(ctx context.Context, cfg *config.Config) error {
+	topologyInfo, err := is.getTopologyInfo(cfg)
+	if err != nil {
+		is.lg.Error("get topology failed", zap.Error(err))
+		return err
+	}
+
+	childCtx, cancelFunc := context.WithCancel(ctx)
+	is.cancelFunc = cancelFunc
+	is.wg.Run(func() {
+		is.updateTopologyLivenessLoop(childCtx, topologyInfo)
+	})
+	return nil
+}
+
+func (is *InfoSyncer) updateTopologyLivenessLoop(ctx context.Context, topologyInfo *TopologyInfo) {
+	// We allow TiProxy to start before PD, so do not fail in the main goroutine.
+	if err := is.initTopologySession(ctx); err != nil {
+		return
+	}
+	_ = is.storeTopologyInfo(ctx, topologyInfo)
+	_ = is.updateTopologyAliveness(ctx, topologyInfo)
+	ticker := time.NewTicker(is.syncConfig.refreshIntvl)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// Even if the topology is unchanged, the server may restart.
+			// We don't assume the server still persists data after restart, so we always store it again.
+			_ = is.storeTopologyInfo(ctx, topologyInfo)
+			_ = is.updateTopologyAliveness(ctx, topologyInfo)
+		case <-is.topologySession.Done():
+			is.lg.Warn("restart topology session")
+			if err := is.initTopologySession(ctx); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (is *InfoSyncer) initTopologySession(ctx context.Context) error {
+	// Infinitely retry until cancelled.
+	return retry.RetryNotify(func() error {
+		topologySession, err := concurrency.NewSession(is.etcdCli, concurrency.WithTTL(is.syncConfig.sessionTTL), concurrency.WithContext(ctx))
+		if err == nil {
+			is.topologySession = topologySession
+			is.lg.Info("topology session is initialized", zap.Error(err))
+		}
+		return err
+	}, ctx, is.syncConfig.putRetryIntvl, retry.InfiniteCnt,
+		func(err error, duration time.Duration) {
+			is.lg.Error("failed to init topology session, retrying", zap.Error(err))
+		}, logInterval)
```

Comment:
Why not do like `storeTopologyInfo`? I mean, only log the return of `retry.RetryNotify()`.

Reply:
Because it's retrying infinitely, until it succeeds or TiProxy quits.
If I log it only when the loop quits, I can't figure out why TiProxy doesn't register itself to TiDB because there's no log yet.